### PR TITLE
LPS-173047 Anonymous users are exported to LDAP

### PR DESCRIPTION
**Pull request by istvansajtos**
Hi Team,

This is the continuation of https://github.com/liferay-usm/liferay-portal/pull/944, forwarding this to AppSec team as requested by @noraszel.

Please let me know if I can help you with anything!

Thank you,
Istvan

Ticket: [LPS-173047](https://issues.liferay.com/browse/LPS-173047)

### CODE REVIEW

**Review by oliverkecskemety — COMMENTED** [review]
Hi @istvansajtos,

Thank you for the PR! Overall looks good to me but please check and consider my questions!

Also please rebase to avoid conflicting files.

File: modules/apps/user-associated-data/user-associated-data-web/src/main/java/com/liferay/user/associated/data/web/internal/anonymizer/UADAnonymousUserProviderImpl.java [hunk 1]
```diff
@@ -76,8 +84,16 @@ public boolean isAnonymousUser(User user) {
 		}
 	}
 
+	private void _completeAnonymousUser(User anonymousUser) throws Exception {
+		_userLocalService.updateStatus(
+			anonymousUser.getUserId(), WorkflowConstants.STATUS_INACTIVE,
+			new ServiceContext());
+	}
```
[the remark this answers is not in the document]
Comment (oliverkecskemety) [reply modules/apps/user-associated-data/user-associated-data-web/src/main/java/com/liferay/user/associated/data/web/internal/anonymizer/UADAnonymousUserProviderImpl.java:91]: Sorry, I found the answer in the ticket description.

File: modules/apps/user-associated-data/user-associated-data-web/src/main/java/com/liferay/user/associated/data/web/internal/anonymizer/UADAnonymousUserProviderImpl.java [hunk 2]
```diff
@@ -106,26 +122,55 @@ private User _createAnonymousUser(long companyId) throws Exception {
...
-			anonymousUser.getUserId(), WorkflowConstants.STATUS_INACTIVE,
-			new ServiceContext());
-
-		return anonymousUser;
+		_userLocalService.addUser(user);
+
+		_groupLocalService.addGroup(
+			user.getUserId(), GroupConstants.DEFAULT_PARENT_GROUP_ID,
+			User.class.getName(), user.getUserId(),
+			GroupConstants.DEFAULT_LIVE_GROUP_ID, (Map<Locale, String>)null,
```
[the remark this answers is not in the document]
Comment (istvansajtos) [reply modules/apps/user-associated-data/user-associated-data-web/src/main/java/com/liferay/user/associated/data/web/internal/anonymizer/UADAnonymousUserProviderImpl.java:147]: Nice catch, @oliverkecskemety ! I copied part of my changes from UserLocalServiceImpl and this also came from there: https://github.com/liferay/liferay-portal/blob/master/portal-impl/src/com/liferay/portal/service/impl/UserLocalServiceImpl.java#L1228-L1233

According to the resources I've found about this practice, this is only for the compiler to be able to choose from overloaded method. After checking GroupLocalService, currently there is no other method with the same parameter count, so it is safe to remove the casting. I am going to update and force push.

**Comment by istvansajtos** [comment]
I've removed the casting of null reference and rebased the commits.

**Comment by istvansajtos** [comment]
ci:test:sf

**Comment by istvansajtos** [comment]
ci:test:relevant

**Comment by istvansajtos** [comment]
ci:test:security

File: modules/apps/user-associated-data/user-associated-data-web/src/main/java/com/liferay/user/associated/data/web/internal/anonymizer/UADAnonymousUserProviderImpl.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (oliverkecskemety) [reply modules/apps/user-associated-data/user-associated-data-web/src/main/java/com/liferay/user/associated/data/web/internal/anonymizer/UADAnonymousUserProviderImpl.java]: Thank you and thanks for the explanation too! I didn't know that.

**Review by oliverkecskemety — APPROVED** [review]

**Comment by oliverkecskemety** [comment]
Thank you @istvansajtos! Everything looks really good to me!

**Comment by oliverkecskemety** [comment]
ci:forward

**Pull request closed without merging**